tests/bsim/bluetooth/host/gatt/authorization/src/gatt_client_test.c

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/gatt.h>
+
+#include "common.h"
+
+CREATE_FLAG(flag_is_connected);
+CREATE_FLAG(flag_discover_complete);
+CREATE_FLAG(flag_write_complete);
+CREATE_FLAG(flag_read_complete);
+
+static struct bt_conn *g_conn;
+static uint16_t unhandled_chrc_handle;
+static uint16_t unauthorized_chrc_handle;
+static uint16_t authorized_chrc_handle;
+static uint16_t cp_chrc_handle;
+static const struct bt_uuid *test_svc_uuid = TEST_SERVICE_UUID;
+
+#define ARRAY_ITEM(i, _) i
+static uint8_t chrc_data[] = { LISTIFY(CHRC_SIZE, ARRAY_ITEM, (,)) }; /* 1, 2, 3 ... */
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err != 0) {
+		FAIL("Failed to connect to %s (%u)\n", addr, err);
+		return;
+	}
+
+	printk("Connected to %s\n", addr);
+
+	g_conn = conn;
```

Comment:
I think you are missing a ref here.

Reply:
@alwa-nordic should this stop the PR from being merged?

Reply:
The reference call seems unnecessary in this case. Please see the documentation of the `bt_conn_le_create` that creates a reference that the API user needs to release using the `bt_conn_unref` (as it is done in the disconnected callback).

```
/** @brief Initiate an LE connection to a remote device.
 *
 *  Allows initiate new LE link to remote peer using its address.
 *
 *  The caller gets a new reference to the connection object which must be
 *  released with bt_conn_unref() once done using the object.
```

FYI, the connection logic in this test project was copied from the following project: `tests/bsim/bluetooth/host/gatt/general`

Reply:
@carlescufi Yes. I think this should be resolved first.

@kapi-no I see. So you already have a pointer value with reference ownership in `g_conn`. It's semantically weird that you overwrite it like that. I can elaborate on this if you want. Can you replace this with `__ASSERT_NO_MSG(g_conn == conn)`?

Reply:
Sure, I can also align it in the source project in a separate PR

Reply:
addressed

Reply:
I also opened a fix for the source project in the following PR:

https://github.com/zephyrproject-rtos/zephyr/pull/67837